Handle non-existing file in Backup password check too

agners opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
Make sure we handle a non-existing backup file also when validating the password.

<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

## Type of change

<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (which adds functionality to the supervisor)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:
- Link to cli pull request:
- Link to client library pull request:

## Checklist

<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Ruff (`ruff format supervisor tests`)
- [ ] Tests have been added to verify that the new code works.

If API endpoints or add-on configuration are added/changed:

- [ ] Documentation added/updated for [developers.home-assistant.io][docs-repository]
- [ ] [CLI][cli-repository] updated (if necessary)
- [ ] [Client library][client-library-repository] updated (if necessary)

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[docs-repository]: https://github.com/home-assistant/developers.home-assistant
[cli-repository]: https://github.com/home-assistant/cli
[client-library-repository]: https://github.com/home-assistant-libs/python-supervisor-client/


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Bug Fixes**
	- Enhanced error handling for backup operations to present clearer messages when backups are missing or cannot be read.
	- Improved the backup password validation process by enabling a smoother recovery path when specific errors occur.
	- Streamlined logic for validating backup locations and passwords to ensure accuracy based on specified locations.

- **New Features**
	- Added a new test function to validate the password for backups, covering various scenarios and edge cases.

- **Tests**
	- Updated test functions to allow dynamic selection of backup files, improving flexibility and robustness of test scenarios.
	- Refined handling of backup instance properties during tests, enhancing clarity and maintainability.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

mdegat01 commented on supervisor/backups/manager.py:
```diff
@@ -727,10 +727,14 @@ async def _validate_location_password(
+                        f"Invalid password for backup {backup.slug}", _LOGGER.error
+                    )
+            except BackupFileNotFoundError:
+                self.sys_create_task(self.sys_backups.reload(location))
```
```suggestion
                self.sys_create_task(self.reload(location))
```
`sys_backups` is how other objects get to backup manager but that's where we are already. Although we could just do this call in the other `except` before raising `BackupFileNotFoundError` and drop this `try` `except` if we prefer, we don't have to schedule the `reload` from here specifically.

agners replied:
We don't have the location in the other try except...

mdegat01 replied:
oh, good point  👍

mdegat01 reviewed (changes requested):
Should probably have a test case as well. Should just be able to add a parameter set of a partial restore with an invalid password to here:
https://github.com/home-assistant/supervisor/blob/4c04f364a3bd265842bbc765fed7a7c2e7e00fa2/tests/api/test_backups.py#L1122-L1163

mdegat01 commented:
@agners in retrospect, our conversation above should've raised warning flags to me. The `validate_password` method needs to know the location, otherwise it cannot check the password correctly. Since we now support having the same backup in mulitiple locations with different encryption options, we cannot validate the password without knowing the location. We were only validating the password using the default location even if that's not where we were restoring from.

I adjusted the code to fix this bug, removed the unnecessary `try` `catch` and added a test case to keep this all working going forward.

agners commented:
> @agners in retrospect, our conversation above should've raised warning flags to me. The `validate_password` method needs to know the location, otherwise it cannot check the password correctly. Since we now support having the same backup in mulitiple locations with different encryption options, we cannot validate the password without knowing the location. We were only validating the password using the default location even if that's not where we were restoring from.

Uh interesting, that could explain some of the issues we have been seeing that the password was not considered valid in some cases maybe :thinking: 


> I adjusted the code to fix this bug, removed the unnecessary `try` `catch` and added a test case to keep this all working going forward.

Cool, thanks! There was some more fallout in not properly mocked test cases (the test cases mocked the `protected` property directly, instead of the new ultimate source of truth in `all_locations`. Tests should be fine now.

agners commented on supervisor/backups/backup.py:
```diff
@@ -433,7 +433,13 @@ def _validate_file() -> bool:
 
-        return await self.sys_run_in_executor(_validate_file)
+        try:
+            await self.sys_run_in_executor(_validate_file)
```
The return statement here was actually missing :see_no_evil: 

I've added a test to cover `validate_file`, with that this should no longer happen.
